dspy/primitives/program.py

```diff
@@ -25,6 +30,99 @@ def __init__(self):
     def __call__(self, *args, **kwargs):
         return self.forward(*args, **kwargs)
 
+    def _is_structurally_equivalent(self, other: Module) -> bool:
+        """ Check if two modules are structurally equivalent by comparing the names and structure of their predictors.
+        
+        Args:
+            other: The module to compare with.
+        
+        Returns:
+            bool: `True` if the modules are structurally equivalent, `False` otherwise.
+        """
+        # Check if the two modules are instances of the Program class
+        if not isinstance(other, self.__class__):
+            return False
+
+        # Check if the two modules have the same number of predictors
+        if len(self.predictors()) != len(other.predictors()):
+            return False
+
+        # Check if the two modules have structurally equivalent predictors
+        for (name1, pred1), (name2, pred2) in zip(self.named_predictors(), other.named_predictors()):
+            if name1 != name2 or not pred1._is_structurally_equivalent(pred2):
+                return False
+
+        return True
+
+    def _has_shared_predictor(self, other: Module) -> bool:
+        """ Check if two modules share a predictor that points to the same Python object.
+        
+        Args:
+            other: The module to compare with.
+        
+        Returns:
+            bool: `True` if the modules share a predictor, `False` otherwise.
+        """
```

Comment:
would be good to add a similar check as L46-L48

Reply:
Ah, this is a great catch, I'll modify the check to ensure there is no predictor overlap irrespective of the order of the predictor...

Reply:
Updated